LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DefaultLoRaDataRequestHandler.cs

```diff
@@ -393,27 +388,58 @@ public async Task<LoRaDeviceRequestProcessResult> ProcessRequestAsync(LoRaReques
             }
         }
 
+        void HandlePreferredGatewayChanges(
+            LoRaRequest request,
+            LoRaDevice loRaDevice,
+            FunctionBundlerResult bundlerResult)
+        {
+            var preferredGatewayResult = bundlerResult.PreferredGatewayResult;
+            if (preferredGatewayResult.IsSuccessful())
+            {
+                var currentIsPreferredGateway = bundlerResult.PreferredGatewayResult.PreferredGatewayID == this.configuration.GatewayID;
+
+                var preferredGatewayChanged = bundlerResult.PreferredGatewayResult.PreferredGatewayID != loRaDevice.PreferredGatewayID;
+                if (preferredGatewayChanged)
+                    Logger.Log(loRaDevice.DevEUI, $"Preferred gateway changed from '{loRaDevice.PreferredGatewayID}' to '{preferredGatewayResult.PreferredGatewayID}'", LogLevel.Information);
+
+                if (preferredGatewayChanged)
+                {
+                    loRaDevice.UpdatePreferredGatewayID(bundlerResult.PreferredGatewayResult.PreferredGatewayID, acceptChanges: !currentIsPreferredGateway);
+                }
+
+                // Save the region if we are the winning gateway and it changed
+                if (request.LoRaRegion.LoRaRegion != loRaDevice.Region)
```

Comment:
LoRaRegion.LoRaRegion? Is that intentional? Could we simplify that and use the region directly? Or change the name of the final field to Region as well?

Reply:
Changed to:
LoRaRegion &rarr; enum defining the regions (eu, us, ...)
Region &rarr; the actual region object, that has a LoRaRegion property as well